src/csharp/Microsoft.Spark/Sql/RowCollector.cs

```diff
@@ -37,5 +38,81 @@ public IEnumerable<Row> Collect(ISocketWrapper socket)
                 }
             }
         }
+
+        /// <summary>
+        /// Synchronously collects pickled row objects from the given socket.
+        /// </summary>
+        /// <param name="socket">Socket the get the stream from</param>
+        /// <param name="server">The jvm socket auth server</param>
+        /// <returns>Collection of row objects</returns>
+        public IEnumerable<Row> SynchronousCollect(ISocketWrapper socket, JvmObjectReference server) =>
+            new SynchronousRowCollector(socket, server);
+
+        /// <summary>
+        /// SynchronousRowCollector synchronously collects Row objects from a socket.
+        /// </summary>
+        private class SynchronousRowCollector : IEnumerable<Row>
+        {
+            private readonly ISocketWrapper _socket;
+            private readonly JvmObjectReference _server;
+
+            private int _readStatus = 1;
+            private IEnumerable<Row> _collectEnumerable = null;
+
+            internal SynchronousRowCollector(ISocketWrapper socket, JvmObjectReference server)
+            {
+                _socket = socket;
+                _server = server;
+            }
+
+            ~SynchronousRowCollector()
```

Comment:
I believe an exception thrown in finalizer is fatal. Should we guard against that?

Reply:
Added try/catch and ignoring the exceptions.